### PR TITLE
[SPARSE] Add support for sparse MKLGPU backend

### PR DESCRIPTION
**Pull request by Rbiessy**
# Description

Add support for sparse MKLGPU backend.
This enables the sparse_blas domain by default.

Depends on https://github.com/oneapi-src/oneMKL/pull/407

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
- [x] Have you formatted the code using clang-format?

## New features

- [x] Have you provided motivation for adding a new feature?
- [x] Have you added relevant tests?

### CODE REVIEW

**Comment by Rbiessy** [comment]
Currently there seems to be an issue somewhere when running ctest on PVC. I'm attaching the logs of the tests and example run separately:
[log_tests_rt_gpu.txt](https://github.com/oneapi-src/oneMKL/files/13390337/log_tests_rt_gpu.txt)
[log_tests_ct_gpu.txt](https://github.com/oneapi-src/oneMKL/files/13390338/log_tests_ct_gpu.txt)
[log_example_gpu.txt](https://github.com/oneapi-src/oneMKL/files/13390339/log_example_gpu.txt)

File: tests/unit_tests/sparse_blas/source/sparse_trsv_buffer.cpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+extern std::vector<sycl::device *> devices;
+
+namespace {
+
+template <typename fpType, typename intType>
+int test(sycl::device *dev, intType m, double density_A_matrix, oneapi::mkl::index_base index,
+         oneapi::mkl::uplo uplo_val, oneapi::mkl::transpose transpose_val,
+         oneapi::mkl::diag diag_val, bool use_optimize) {
+    sycl::queue main_queue(*dev, exception_handler_t());
```
Comment (gajanan-choudhary) [inline tests/unit_tests/sparse_blas/source/sparse_trsv_buffer.cpp]: Are the test files between CPU and GPU, or other backends for that matter, all the same? If yes, that's great!

Comment (Rbiessy) [reply tests/unit_tests/sparse_blas/source/sparse_trsv_buffer.cpp]: Yes the tests are backend agnostic! This is nice but can be an issue when some backends have different limitations.

**Comment by gajanan-choudhary** [comment]
Once you are done merging the approved PR #407, could you please rebase and post the new test logs (with the --terse-output flag)?

**Comment by Rbiessy** [comment]
@gajanan-choudhary 
Logs using oneAPI 2024.0 on iGPU and Arc: 
[log_sparse_mklgpu_oneapi_2024_0.txt](https://github.com/oneapi-src/oneMKL/files/13513714/log_sparse_mklgpu_oneapi_2024_0.txt)
Logs on PVC using a recent nightly:
[log_sparse_mklgpu_nightly_pvc_ct.txt](https://github.com/oneapi-src/oneMKL/files/13513722/log_sparse_mklgpu_nightly_pvc_ct.txt)
[log_sparse_mklgpu_nightly_pvc_rt.txt](https://github.com/oneapi-src/oneMKL/files/13513725/log_sparse_mklgpu_nightly_pvc_rt.txt)
[log_sparse_mklgpu_nightly_pvc_example.txt](https://github.com/oneapi-src/oneMKL/files/13513726/log_sparse_mklgpu_nightly_pvc_example.txt)
Configurations using gemm with `transpose_B != nontrans` and trsv with `transpose_val != nontrans` are skipped.

File: src/sparse_blas/backends/mklgpu/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+target_include_directories(${LIB_OBJ}
+  PRIVATE ${PROJECT_SOURCE_DIR}/include
+          ${PROJECT_SOURCE_DIR}/src
+          ${CMAKE_BINARY_DIR}/bin
+          ${ONEMKL_GENERATED_INCLUDE_PATH}
+)
+
+target_compile_options(${LIB_OBJ} PRIVATE ${ONEMKL_BUILD_COPT})
+if (USE_ADD_SYCL_TO_TARGET_INTEGRATION)
+  add_sycl_to_target(TARGET ${LIB_OBJ} SOURCES ${SOURCES})
```
Comment (FMarno) [inline src/sparse_blas/backends/mklgpu/CMakeLists.txt]: Where is `SOURCES` defined?

**Review by FMarno — COMMENTED** [review]
LGTM

File: src/sparse_blas/backends/mklgpu/CMakeLists.txt (hunk 2 above)
Comment (Rbiessy) [reply src/sparse_blas/backends/mklgpu/CMakeLists.txt]: Thank you, it was not defined so empty. This is just leftovers from other backends that use this. We don't need this for sparse_blas so I have removed it in the CPU and GPU backends.
We have the same "issue" in the DFT domain btw.

**Review by hjabird — APPROVED** [review]
LGTM

File: src/sparse_blas/backends/mklgpu/CMakeLists.txt (hunk 2 above)
Comment (FMarno) [reply src/sparse_blas/backends/mklgpu/CMakeLists.txt]: I also checked the DFT domain and saw that :sweat_smile:

**Review by FMarno — APPROVED** [review]

**Comment by gajanan-choudhary** [comment]
> @gajanan-choudhary Logs using oneAPI 2024.0 on iGPU and Arc: [log_sparse_mklgpu_oneapi_2024_0.txt](https://github.com/oneapi-src/oneMKL/files/13513714/log_sparse_mklgpu_oneapi_2024_0.txt) Logs on PVC using a recent nightly: [log_sparse_mklgpu_nightly_pvc_ct.txt](https://github.com/oneapi-src/oneMKL/files/13513722/log_sparse_mklgpu_nightly_pvc_ct.txt) [log_sparse_mklgpu_nightly_pvc_rt.txt](https://github.com/oneapi-src/oneMKL/files/13513725/log_sparse_mklgpu_nightly_pvc_rt.txt) [log_sparse_mklgpu_nightly_pvc_example.txt](https://github.com/oneapi-src/oneMKL/files/13513726/log_sparse_mklgpu_nightly_pvc_example.txt) Configurations using gemm with `transpose_B != nontrans` and trsv with `transpose_val != nontrans` are skipped.

In [log_sparse_mklgpu_oneapi_2024_0.txt](https://github.com/oneapi-src/oneMKL/files/13513714/log_sparse_mklgpu_oneapi_2024_0.txt) I see:
```
67/97 Test #67: SPARSE_BLAS/CT/SparseGemvBufferTestSuite/SparseGemvBufferTests.RealDoublePrecision/Intel_R__Arc_TM__A770_Graphics ......***Skipped   0.11 sec
      Start 68: SPARSE_BLAS/CT/SparseGemvBufferTestSuite/SparseGemvBufferTests.RealDoublePrecision/Intel_R__UHD_Graphics_750
```
Why are some `gemv` tests (presumably subtests) being reported as being skipped? I believe GEMV is supported in all combinations in oneMKL: cpu/gpu, buffer/usm, int32/int64, float/double/complex<float>/complex<double>, and nontrans/trans/conjtrans. Are we unintentionally throwing an `unimplemented` exception somewhere in GEMV?

**Comment by gajanan-choudhary** [comment]
> > @gajanan-choudhary Logs using oneAPI 2024.0 on iGPU and Arc: [log_sparse_mklgpu_oneapi_2024_0.txt](https://github.com/oneapi-src/oneMKL/files/13513714/log_sparse_mklgpu_oneapi_2024_0.txt) Logs on PVC using a recent nightly: [log_sparse_mklgpu_nightly_pvc_ct.txt](https://github.com/oneapi-src/oneMKL/files/13513722/log_sparse_mklgpu_nightly_pvc_ct.txt) [log_sparse_mklgpu_nightly_pvc_rt.txt](https://github.com/oneapi-src/oneMKL/files/13513725/log_sparse_mklgpu_nightly_pvc_rt.txt) [log_sparse_mklgpu_nightly_pvc_example.txt](https://github.com/oneapi-src/oneMKL/files/13513726/log_sparse_mklgpu_nightly_pvc_example.txt) Configurations using gemm with `transpose_B != nontrans` and trsv with `transpose_val != nontrans` are skipped.
> 
> In [log_sparse_mklgpu_oneapi_2024_0.txt](https://github.com/oneapi-src/oneMKL/files/13513714/log_sparse_mklgpu_oneapi_2024_0.txt) I see:
> 
> ```
> 67/97 Test #67: SPARSE_BLAS/CT/SparseGemvBufferTestSuite/SparseGemvBufferTests.RealDoublePrecision/Intel_R__Arc_TM__A770_Graphics ......***Skipped   0.11 sec
>       Start 68: SPARSE_BLAS/CT/SparseGemvBufferTestSuite/SparseGemvBufferTests.RealDoublePrecision/Intel_R__UHD_Graphics_750
> ```
> 
> Why are some `gemv` tests (presumably subtests) being reported as being skipped? I believe GEMV is supported in all combinations in oneMKL: cpu/gpu, buffer/usm, int32/int64, float/double/complex/complex, and nontrans/trans/conjtrans. Are we unintentionally throwing an `unimplemented` exception somewhere in GEMV?

Nevermind, I just realized the skipped tests are on "Arc_TM__A770_Graphics", which only supports single precision, so all double precision tests get skipped on that hardware. On PVC, the other log files show all GEMV tests to be passing.

**Review by gajanan-choudhary — APPROVED** [review]
LGTM. I suppose I have an AR to now to work with Infra to enable regular testing of the sparse BLAS domain in this repository.

**Pull request merged**